### DISCUSSION TITLE
Custom Action "Download PDF" button

### DISCUSSION THREAD

**Discussion by DevMushref** [filamentphp/filament#5289] [Help] [answered]
I am using this package: https://github.com/elibyy/tcpdf-laravel which is a wrapper for the original https://tcpdf.org/

I was using Metronic 8 for the project and after discovering Filament, I decided to give it a try and I'm very close to migrate my project into Filament.

In the current project (Not Filament) I have 4 actions in the table (View, Edit, Delete, Download PDF):

![image](https://user-images.githubusercontent.com/25293489/208400801-fc86ec90-d809-4e4c-a2b1-3c946844a25f.png)

The PDF button is handed by TCPDF to convert inserted data to a PDF report once clicked ,like so:

![Screenshot 2022-12-19 131030](https://user-images.githubusercontent.com/25293489/208401543-c5ec64f8-5119-4cab-984b-16b59f9378f5.png)

The content of the pdf button is a separate blade file called from the controller, so you may consider it as a "Custom HTML"

How can I implement this action button in Filament? I didn't find a "Custom Actions" sections in the docs.

Note that the PDF button should be within the View page also.

Thank you

**Comment by ArtMin96** [comment]
```php

use Filament\Tables\Actions\Action;

class DownloadPdfAction extends Action
{
    protected string $view = 'actions.download-pdf-view';

    public static function make(?string $name = 'download-pdf'): static
    {
        return parent::make($name);
    }

    protected function setUp(): void
    {
        //...
    }
}

**Comment by danharrin** [comment]
You could use Livewire's file downloads

```php
Action::make('download')
    ->label('Download PDF')
    ->action(function ($record) {
        // generate PDF here from the current $record
        
        return response()->streamDownload(...);
    })
```

https://laravel-livewire.com/docs/2.x/file-downloads#introduction

**Reply by DevMushref** [reply]
When I try to add PDF class like the TCPDF docs withing the `->action()` I get an error "Undefined type, PDF"

```
->actions([
                Tables\Actions\ViewAction::make(),
                Tables\Actions\EditAction::make(),
                Tables\Actions\Action::make('download')
                    ->label('PDF')
                    ->action(function ($record) {
                        // Generate PDF here from the current $record
                        PDF::SetTitle('IR#');
                        PDF::AddPage();
                        PDF::SetMargins(20, 20, 20);
                        PDF::Output(public_path('test.pdf'), 'I');
                        PDF::reset();
                    }),
            ])
```

**Reply by ArtMin96** [reply]
Maybe that problem comes from PDF class or somewhere else? Can you send the whole exception?

**Reply by DevMushref** [reply]
The issue is I'm not getting any exceptions, the error is only shown in VSCode itself...

![image](https://user-images.githubusercontent.com/25293489/208836126-d726b349-1779-4585-b5f2-24c6f21f239a.png)

The button is shown in the table, but when clicked it's in a loading loop.

**Reply by ArtMin96** [reply]
That error is not related to the Filament.

**Reply by danharrin** [reply]
You didnt import the PDF class

**Comment by DevMushref** [comment]
I decided to change TCPDF to domPDF as it's a lot easier (TCPDF is develop on top of domPDF anyways).

```
Tables\Actions\Action::make('download')
                    ->label('PDF')
                    ->action(function ($record) {
                        return Pdf::loadView('pdf.report_pdf', $record)->download('test.pdf');
                    }),
```

The code above brings me this error:

![image](https://user-images.githubusercontent.com/25293489/209776069-4a9edab9-6adc-468f-908e-b38da476c7be.png)

When I `dd($record)` it returns the Model and not the data.

![image](https://user-images.githubusercontent.com/25293489/209776431-0a7f5b58-2073-404d-8d90-16d12431d292.png)

What's the variable that returns the data? I couldn't find what I'm looking for in the documentation. Pointing me to it would be highly appreciated also.

**Reply by ArtMin96** [reply]
```php
->action(function(array $data) {
    //...
})
```

**Reply by DevMushref** [reply]
![image](https://user-images.githubusercontent.com/25293489/209779766-88a4ff4d-1471-4d02-b301-bcca3eea89fa.png)

why is it returning empty array?

**Reply by ArtMin96** [reply]
Sorry for the BulkAction, the way I said is wrong. Try something like 
```php
->action(function(YourModel $record) {
    return Pdf::loadView('pdf.report_pdf', $record->toArray())->download('test.pdf');
})

**Comment by DevMushref** [comment, marked as answer]
### Got it!!

Instead of using `->action()`, use `->url()`

domPDF needs a controller and a route in order to generate PDFs correctly.

### **First off, create a route:**

```
Route::prefix('generate-pdf')->name('generate-pdf.')
    ->group(function () {
        Route::controller(generatePDF::class)->group(function () {
            Route::get('incident-report/{record}', 'incidentReport')->name('incident.report'); // Incident Reports
        });
    });
```

I created a route group called `generate-pdf` so within it I will create routes that redirects to the controller which contains methods of generating different PDF files based on Filament resource. In my case I only have 1 for now which is: 

`Route::get('incident-report/{record}', 'incidentReport')->name('incident.report'); // Incident Reports`

This route is handled by the controller method:

```
public function incidentReport($id)
    {
        $record = IncidentReport::find($id);
        $pdf = App::make('dompdf.wrapper');
        $pdf->loadView('pdf.report_pdf', compact('record')); // Pass the variable $record to the blade file
        return $pdf->stream(); // renders the PDF in the browser
    }
```

### Moving to Filament Resource

1. Create the action button.
2. set the Url, make it open in a new tab (if wanted).
3. Inside the url, set the route you created in `web.php`

```
->actions([
                Tables\Actions\Action::make('download')
                ->label('PDF')
                ->url(
                    fn (IncidentReport $record): string => route('generate-pdf.incident.report', ['record' => $record]),
                    shouldOpenInNewTab: true
                )
            ])
```

This way, the pdf generates as I want.

![ezgif com-gif-maker](https://user-images.githubusercontent.com/25293489/209818022-e943794f-c1f1-41fa-a89f-758a4c06dce7.gif)

**Reply by elysdan** [reply]
Hi bro, I need your help about this, you can share with me your code? I try the same that you explain on this post, but not working for me T_T... for example, when I try make actions and test it, appear this error: Maximum execution time of 60 seconds exceeded... about the code controller pdf, when I test, in my view I call what is supposed to be my object, and it does not bring me the value, it gives me the following error: Attempt to read property "codigo" on bool...

![Captura desde 2023-11-24 16-40-33](https://github.com/filamentphp/filament/assets/49619717/e4848a9d-1654-4831-968a-815c01414f55)

![imagen](https://github.com/filamentphp/filament/assets/49619717/49cef1d5-5c47-49e3-83d7-a745e6f57731)

![imagen](https://github.com/filamentphp/filament/assets/49619717/990bfb8a-012d-4612-8eac-419399933580)

**Reply by elysdan** [reply]
I forgot to explain the images, I have already done several different things and it still doesn't give me the expected result.

**Comment by syed1980eer** [comment]
Hello friend.

I have a question related to this, please allow me. In my filament project, I want to create a PDF after filtering the data. Ex. let's assume I have all together 10 order entries out of which 4 got delivered, now when I select 'Delivered' I want to create or download a pdf with delivered entries, not all the 10 records. I'm using DomPDF.

**Reply by KhairulAzmi21** [reply]
Did you manage to find possible solution ?

**Comment by KhairulAzmi21** [comment]
How do we send filter request to controller method ?

**Comment by ramirobrescia** [comment]
Hi I want to share my result inpired by this discussion.

Then, the same concept but reusable:

```php
// Action
class PurchaseReportAction extends Action
{
    use CanCustomizeProcess;

    public static function getDefaultName(): ?string
    {
        return 'purchase report';
    }

    protected function setUp(): void
    {
        parent::setUp();

        $this->label(__('purchase.report.provider.action.label'));
        
        $this->icon('heroicon-o-document-text');

        $this->url(fn (Purchase $record) => route('report.purchase', ['id' => $record->id], true));
    }
}

// Usage in a Resource
public static function table(Table $table): Table
{
    return $table
        ->columns([
           // ...
        ])
        ->filters([
            // ...
        ])
        ->actions([
            Tables\Actions\ActionGroup::make([
                PurchaseReportAction::make(),
                Tables\Actions\EditAction::make(),
            ])->iconButton()
        ])
        // ..
        );
}
```
Thank you! A hug.
